extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientUtils.java

```diff
@@ -4,15 +4,31 @@
 
 import org.eclipse.microprofile.config.ConfigProvider;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.ConfigBuilder;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.utils.KubernetesSerialization;
 import io.quarkus.runtime.TlsConfig;
 
 public class KubernetesClientUtils {
 
     private static final String PREFIX = "quarkus.kubernetes-client.";
+    private static final KubernetesSerialization SERIALIZATION = new KubernetesSerialization(new ObjectMapper(), true);
```

Comment:
Shouldn't we use the Quarkus ObjectMapper here, in case it has been customised by users?

Reply:
This is an important point.

I'd first ask how were users customizing the Fabric8's serialization purposes `ObjectMapper` before.
If mutating `Serialization.jsonMapper()` is the answer to the previous question and this was an extended practice, then we should definitely provide a way to customize this mapper. With the current changes, this will no longer be possible.

The next point would be to know if we want to share a same `ObjectMapper`  used application-wise with the client or have an exclusive one for it. Some mapper configurations might not be ideal for the client or even break it.

Reply:
IMHO, it should not use the user customizable mapper, because the client itself knows everything about the model serialization/deserialization and should not be affected by the user's configuration

Reply:
@geoand this isn't quite true, for example, if people want to use Kotlin or other libs in conjunction to the Fabric8 client, they would need to customise the mapper to make it work correctly with the client.

Reply:
Regarding, @manusa's question, I really don't know how people usually do it or even how we think they should do it. We added a specific customiser in QOSDK to address the problem that people do need to customise the client's mapper for their use case, see [KubernetesClientSerializationCustomizer](https://github.com/quarkiverse/quarkus-operator-sdk/blob/aa78e13579d8eb958474e319d91dada71e3acdf9/core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/KubernetesClientSerializationCustomizer.java), which is meant to be used as a qualifier on a Quarkus `ObjectMapperCustomizer` specifically to customise the Fabric8 mapper.

Reply:
And yes, we do currently retrieve the Fabric8's mapper via `Serialisation.jsonMapper()`. This is how we currently customise the mapper: https://github.com/quarkiverse/quarkus-operator-sdk/blob/89a95cd489c46c5a17660faa0f82b7cdbbd7049e/core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/ConfigurationServiceRecorder.java#L69-L77.
I'm open to changes provided they can happen before 3.2.

Reply:
Maybe should the client expose a producer method so that interested third party could pick up the mapper via CDI injection?

Reply:
I guess the core question is whether or not there should only be one instance of that mapper for all client instances or if that's not expected, in which case, I think we should probably document how people are supposed to interact with the mapper and what the expectations should be, in particular when someone needs to customise that instance for their needs, because one usually need to share the mapper between the client and the rest of their code when dealing with the kubernetes client (as you probably don't want to have to deal with issues rising from differently configured mappers in the same code base).

Reply:
The way I see it, the Quarkus object mapper is useful for serialization / deserialization where the user has control over the model being exchanged.
In the k8s case, that certainly does not apply so it does not not make sense IMO to allow users to shoot themselves in the foot.

In my view, the JOSDK should provide a CDI bean that is used to configure the client - that bean should provider whatever configuration is necessary to enable the niche use cases that require something other than the default mapper.

Reply:
>  In the k8s case, that certainly does not apply so it does not not make sense IMO to allow users to shoot themselves in the foot.

Again, that is not correct. If you want to use kotlin with the fabric8 client, you will need to customise the mapper (at least, as far as I understand it). If you can't customise the mapper, you won't be able to use the client in many cases. Another use case is when your code needs to interact with both the client and third party libraries. For example, one of our users needs to deal with CloudEvents, which requires the mapper to be customised with a module in order to properly work. Are they expected to maintain another mapper to deal with that part when they also need to serialise Kubernetes objects, sometimes within the same object hierarchy (i.e. they might need to serialise parts of a Kubernetes resource in the context of a CloudEvent instance).
 
> In my view, the JOSDK should provide a CDI bean that is used to configure the client - that bean should provider whatever configuration is necessary to enable the niche use cases that require something other than the default mapper.

I don't think that these issues are limited to JOSDK, though. The examples I gave above certainly aren't JOSDK-specific.

Reply:
> Again, that is not correct. If you want to use kotlin with the fabric8 client, you will need to customise the mapper (at least, as far as I understand it).

Again, this is a very specific use case that should not require users to customize the default mapper.

> Another use case is when your code needs to interact with both the client and third party libraries. For example, one of our users needs to deal with CloudEvents, which requires the mapper to be customised with a module in order to properly work. Are they expected to maintain another mapper to deal with that part when they also need to serialise Kubernetes objects, sometimes within the same object hierarchy (i.e. they might need to serialise parts of a Kubernetes resource in the context of a CloudEvent instance)

If they don't they are asking for trouble, plain and simple.

Reply:
As far as I'm aware, without being a kotlin expert, you cannot use the fabric8 client from kotlin without customizing the fabric8 mapper. Similarly, as far as I'm aware, you cannot have two different mappers take part into the serialisation of one single object. So I would very much like to hear your actual solutions to both of these issues apart from saying you shouldn't customise the mapper because that doesn't really get our users anywhere.

Reply:
I created a bean and qualifier to be able to inject the KubernetesClient-specific ObjectMapper for further modifications such as the Kotlin use case. (see `@KubernetesClientObjectMapper`)

Reply:
Wouldn't it make more sense to implement an approach similar to what I did in QOSDK, i.e. using the customizer approach, instead of this approach which would require people to create a whole new `ObjectMapper` (at least, if I understand things properly)? That seems like exactly the kind of thing we would rather avoid because that would require people to know how the original mapper is configured so that they don't break the client's functionality. We would rather want people to add their configuration on top of what is already configured instead of doing it from scratch.

Also, with this approach, shouldn't the producing method of `KubernetesClientObjectMapperProducer` be marked as `DefaultBean`?

Reply:
For your approach yes.
When I did this I was actually thinking of the opposite flow, injecto to customize, but it does make more sense for users to maybe provide their own at this point. Will change this.

Reply:
As explained above customisation could occur for a variety of reasons. I still think that users should customise the instance rather than provide their own. We might not use any customisation at this point, but should that change, users would need to use the same customisation and add theirs on top so I really believe customizing the existing instance is a better choice than injecting a new one. This would also allows us to control at the customisation point that the customisations are indeed compatible with the needs of the client.